### PR TITLE
Df attributes

### PR DESCRIPTION
**Pull request by dflanderka**

### CODE REVIEW

File: src/input/type_base.cc [hunk 1]
```diff
@@ -28,6 +28,7 @@
 #include "type_record.hh"
 #include "type_output.hh"
 #include "type_repository.hh"
+#include "json_to_storage.hh"
```
Comment (jbrezmorf) [inline src/input/type_base.cc]: Rather just include of JSON parser, i.e. "json_spirit/json_spirit.h".

File: src/input/type_base.hh [hunk 2]
```diff
@@ -154,6 +156,12 @@ public:
      */
     virtual TypeHash content_hash() const =0;
 
+    /// Add attribute to map
+    void add_attribute(std::string name, json_string val);
+
+    /// Print JSON output of attributes to @p stream.
+    void print_json(ostream& stream);
```
Comment (jbrezmorf) [inline src/input/type_base.hh]: Rename to "write_attributes".

File: src/input/type_output.cc [hunk 3]
```diff
@@ -1293,7 +1293,8 @@ void OutputJSONMachine::print_impl(ostream& stream, const Record *type, unsigned
     stream << "\"input_type\" : \"Record\"," << endl;
     stream << "\"type_name\" : \"" << type->type_name() << "\"," << endl;
     stream << "\"type_full_name\" : \"" << type->full_type_name() << "\"," << endl;
-    stream << endl;
+    const_cast<Record *>(type)->print_json(stream);
```
Comment (jbrezmorf) [inline src/input/type_output.cc]: Avoid this const_cast by making print_json const method.

File: src/input/type_base.cc [hunk 4]
```diff
@@ -44,11 +45,13 @@ using namespace std;
 
 
 
-TypeBase::TypeBase() {}
+TypeBase::TypeBase()
+: attributes_( boost::make_shared<attribute_map>() ) {}
```
Comment (jbrezmorf) [inline src/input/type_base.cc]: Not necessary, this is automatic initiation of attributes_.

File: src/input/type_record.cc [hunk 5]
```diff
@@ -475,6 +476,7 @@ RECORD_DECLARE_KEY(AdHocAbstractRecord);
 AbstractRecord::AbstractRecord()
 : child_data_( boost::make_shared<ChildData>( "EmptyAbstractRecord", "" ) )
 {
+	attributes_ = boost::make_shared<attribute_map>();
```
Comment (jbrezmorf) [inline src/input/type_record.cc]: Not necessary. delete.

File: src/input/type_record.cc [hunk 6]
```diff
@@ -483,13 +485,17 @@ AbstractRecord::AbstractRecord()
 
 AbstractRecord::AbstractRecord(const AbstractRecord& other)
 : child_data_(other.child_data_)
-{}
+{
+	attributes_ = other.attributes_;
```
Comment (jbrezmorf) [inline src/input/type_record.cc]: Rather call TypeBase copy constructor in initializer list. Need also add to other copy constructors, are there any for Record or Selection?

**Comment by jbrezmorf** [comment]
Proposed few simplifications. Possible missing call of TypeBase copy constructor for derived classes has to be added.

File: src/input/type_base.cc (hunk 4 above)
Comment (dflanderka) [inline src/input/type_base.cc]: It can't be removed, under cygwin throws error.

Comment (jbrezmorf) [inline src/input/type_base.cc]: Of course. I overlooked the shared pointer.

**Pull request merged**